File: graylog2-server/src/main/java/org/graylog2/alerts/MessageFormatter.java
/**
 * This file is part of Graylog2.
 *
 * Graylog2 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Graylog2 is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Graylog2.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.graylog2.alerts;

import com.google.common.collect.Lists;
import org.graylog2.plugin.Message;

import java.util.*;

/**
 * @author Dennis Oelkers <dennis@torch.sh>
 */
public class MessageFormatter {
    public String formatForMail(Message message) {
        StringBuilder sb = new StringBuilder();

        sb.append("<< Message: ").append(message.getId()).append(" >>\n");
        sb.append("timestamp: ").append(message.getField("timestamp").toString()).append("\n");
        sb.append("source: ").append(message.getSource()).append("\n");

        Map<String, Object> fields = new HashMap<>(message.getFields());
        fields.remove("timestamp");
        fields.remove("source");
        fields.remove("_id");
        String fullMessage = fields.get("message").toString();
        fields.remove("message");

        List<String> fieldKeys = Lists.newArrayList(fields.keySet());
        Collections.sort(fieldKeys, new Comparator<String>() {
            @Override
            public int compare(String o1, String o2) {
                return o1.compareTo(o2);
            }
        });

        for (String key : fieldKeys) {
            sb.append(key)
                    .append(": ")
                    .append(fields.get(key).toString())
                    .append("\n");
        }

        sb.append("message: ").append(fullMessage).append("\n");

        return sb.toString();
    }
}

File: graylog2-web-interface/src/logic/message/MessageFormatter.ts
/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import moment from 'moment';

import MessageFieldsFilter from 'logic/message/MessageFieldsFilter';

type MessageFields = {
  [key: string]: unknown;
  gl2_source_node?: string;
  gl2_source_input?: string;
  streams?: Array<string>;
  timestamp?: string;
};

export type FormattedMessage = {
  id: string;
  timestamp: number;
  filtered_fields: { [key: string]: any };
  formatted_fields: { [key: string]: any };
  fields: { [key: string]: any };
  index: string;
  source_node_id?: string;
  source_input_id?: string;
  stream_ids?: Array<string>;
  highlight_ranges?: { [key: string]: any };
  decoration_stats?: {
    added_fields: { [key: string]: any };
    changed_fields: { [key: string]: any };
    removed_fields: { [key: string]: any };
  };
  [key: string]: any;
};

const MessageFormatter = {
  formatMessageSummary(messageSummary: any): FormattedMessage {
    const { message } = messageSummary;

    return this.formatMessage(
      message._id,
      messageSummary.index,
      message,
      message,
      messageSummary.highlight_ranges,
      messageSummary.decoration_stats,
    );
  },

  formatResultMessage(resultMessage: any): FormattedMessage {
    const { message } = resultMessage;

    return this.formatMessage(
      message.id,
      resultMessage.index,
      message,
      message.fields,
      resultMessage.highlight_ranges,
      resultMessage.decoration_stats,
    );
  },

  formatMessage(id: string, index: string, message: MessageFields, fields: MessageFields, highlightRanges: any, decorationStats: any): FormattedMessage {
    const filteredFields = MessageFieldsFilter.filterFields(fields);

    return {
      id: id,
      timestamp: moment(message.timestamp).unix(),
      filtered_fields: filteredFields,
      formatted_fields: filteredFields,
      fields: fields,
      index: index,
      source_node_id: fields.gl2_source_node,
      source_input_id: fields.gl2_source_input,
      stream_ids: message.streams,
      highlight_ranges: highlightRanges,
      decoration_stats: decorationStats,
    };
  },
};

export default MessageFormatter;
